Remove support schema flags, only use config file. (#2221)
cortexproject/cortex: docs/configuration/schema-config-reference.md

---
title: "Schema Configuration"
linkTitle: "Schema Configuration"
weight: 1
slug: schema-configuration
---

> Cortex uses a NoSQL Store to store its index and optionally an Object store to store its chunks. Cortex has overtime evolved its schema to be more optimal and better fit the use-cases that arised. Currently there are 10 schemas that are used in production. You can move from one schema to another if a new schema fits your purpose better, but you still need to configure Cortex to make sure it can read the old data in the old schemas.

Reviewer: We actually have 9 schemas now.

> Cortex uses a NoSQL Store to store its index and optionally an Object store to store its chunks. Cortex has overtime evolved its schema to be more optimal and better fit the use-cases that arose. Currently there are 9 schemas that are used in production but we recommend running with `v9` schema when possible. You can move from one schema to another if a new schema fits your purpose better, but you still need to configure Cortex to make sure it can read the old data in the old schemas.

Reviewer 2: I would make it clear this applies to the chunks storage engine only.

> Cortex uses a NoSQL Store to store its index and optionally an Object store to store its chunks. Cortex has overtime evolved its schema to be more optimal and better fit the use-cases that arose. Currently there are 9 schemas that are used in production but we recommend running with `v9` schema when possible. You can move from one schema to another if a new schema fits your purpose better, but you still need to configure Cortex to make sure it can read the old data in the old schemas.

Reviewer 2: ```suggestion
Cortex uses a NoSQL Store to store its index and optionally an Object store to store its chunks. Cortex has overtime evolved its schema to be more optimal and better fit the use cases that arose.

Currently there are 9 schema versions that are used in production but we recommend running with `v9` schema when possible. You can move from one schema to another if a new schema fits your purpose better, but you still need to configure Cortex to make sure it can read the old data stored with the old schemas.
```

You can configure the schemas using a YAML config file, that you can point to using the `-schema-config-file` flag. It has the following YAML spec:

```yaml
configs: []<period_config>
```

Where `period_config` is
```

> from: <Date>

Reviewer 2: I would suggest to use the same config syntax of the config file reference, to keep consistency.

store: <string> // The index client to use.
object_store: <string> // The object client to use. If none is specified, `store` is used for storing chunks as well.

> schema: <string> // The schema version to use. Valid ones are v1, v2, v3,..., v11

Reviewer: We don't have `v7` and `v8` schemas.

index: <periodic_table_config>
chunks: <periodic_table_config>
```

Where `periodic_table_config` is
```
prefix: <string> // The prefix to use for the tables.

> period: <time.Duration> // The period at which new tables are created and used. Typically 168h

Reviewer: `The period at which new tables are created and used.` sounds a little confusing for a new user. Any other way to describe it? Maybe let us not correlate with when tables are created and used?

Author: Is it better now?

Reviewer: Much better, just add a space in `theperiod`. Thanks!

tags: <map[string]string> // The tags that can be set on the dynamo table. 
```

Now an example of this file (also something recommended when starting out) is:
```
configs:
  - from: "2020-03-01" // Or typically a week before the Cortex cluster was created.
    schema: v9
    index:
      period: 168h
      prefix: cortex_index_
    chunks:                     // Chunks section is optional and required only if you're not using a separate object store.
      period: 168h
      prefix: cortex_chunks  
    store: aws-dynamo/bigtable-hashed/cassandra/boltdb
    object_store: <above options>/s3/gcs/azure/filesystem
```

An example of an advanced schema file with a lot of changes:
```
configs:

>   - from: "2018-08-23"

Reviewer 2: For each config block I would add a `#` comment to explain what it does. For example, I'm adding a suggestion for the first block:

```suggestion
  # Starting from 2018-08-23 Cortex should store chunks and indexes
  # on Google BigTable using weekly periodic tables. The chunks table
  # names will be prefixed with "dev_chunks_", while index tables will be
  # prefixed with "dev_index_".
  - from: "2018-08-23"
```

    schema: v9
    chunks:
        period: 168h0m0s
        prefix: dev_chunks_
    index:
        period: 168h0m0s
        prefix: dev_index_
    store: gcp-columnkey
  - from: "2019-02-13"
    schema: v9
    chunks:
        period: 168h
        prefix: dev_chunks_
    index:
        period: 168h
        prefix: dev_index_
    object_store: gcs
    store: gcp-columnkey
  - from: "2019-02-24"
    schema: v9
    chunks:
        period: 168h
        prefix: dev_chunks_
    index:
        period: 168h
        prefix: dev_index_
    object_store: gcs
    store: bigtable-hashed
  - from: "2019-03-05"
    schema: v10
    chunks:
        period: 168h
        prefix: dev_chunks_
    index:
        period: 168h
        prefix: dev_index_
    object_store: gcs
    store: bigtable-hashed
```

Note how we started out with v9 and just Bigtable, but later added GCS as the object store, finally moving to v10. This is a complex schema file from out dev cluster that saw a lot of changes dating all the way back to 2018, a typical schema config file usually has one or two schema versions.

### Migrating from flags to schema file

> We've recently removed the ability to configure schema via flags. If you were using flags, you need to port your config from flags to the config file.

Reviewer 2: I would suggest to never use time-sensitive words like "recently" in the doc, because you never know when this doc will be updated again and the "recently" may easily become months or years.

```suggestion
Legacy versions of Cortex did support the ability to configure schema via flags. If you are still using flags, you need to migrate your configuration from flags to the config file.
```

> If you're using:

Reviewer 2: Please add a new line after this, in order to render the markdown list correctly.
